src/common/container_helpers.h

```diff
@@ -0,0 +1,121 @@
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Miscellaneous container helpers.
+
+#pragma once
+
+//local headers
+
+//third party headers
+
+//standard headers
+#include <algorithm>
+#include <functional>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+//forward declarations
+
+
+namespace tools
+{
+
+/// convert a binary comparison function to a functor
+template <typename T, typename ComparisonOpT = bool(const T &a, const T &b)>
+inline auto compare_func(const ComparisonOpT &comparison_function)
+{
+    return [&comparison_function](const T &a, const T &b) -> bool { return comparison_function(a, b); };
+}
+/// test if a container is sorted and unique according to a comparison criteria (defaults to operator<)
+/// NOTE: ComparisonOpT must establish 'strict weak ordering' https://en.cppreference.com/w/cpp/named_req/Compare
+template <typename T, typename ComparisonOpT = std::less<typename T::value_type>>
+bool is_sorted_and_unique(const T &container, const ComparisonOpT &ComparisonOp = ComparisonOpT{})
+{
+    if (!std::is_sorted(container.begin(), container.end(), ComparisonOp))
+        return false;
+
+    if (std::adjacent_find(container.begin(),
+                container.end(),
+                [&ComparisonOp](const typename T::value_type &a, const typename T::value_type &b) -> bool
+                {
+                    return !ComparisonOp(a, b) && !ComparisonOp(b, a);
+                })
+            != container.end())
+        return false;
+
+    return true;
+}
+/// specialization for raw function pointers
+template <typename T>
+bool is_sorted_and_unique(const T &container,
```

Comment:
I missed this earlier, no overload is neeeded for raw functions if the argument is taken by-value.

Reply:
Except for the problem of overload resolution.

Reply:
My frustration with these solutions for compare function overloads - plenty of examples of people using generic predicates don't run into this issue by: (1) using unique function names, (2) using the `struct` solution given in another comment, or (3) using `std::bind` like so: `std::bind<bool(int, int)>(my_compare, _1, _2)`.

I'm not completely against these custom solutions (although I'm wondering suspiciously why solution (1) above isn't being used), but ALSO why not use the `compare_func` above instead of a custom `is_sorted_and_unique` which only works with `const&` arguments for the compare function?

--------------

I also wanted to document (somewhere) this more generic solution I came up with:

```c++
template <typename F>
inline auto select_func(F f)
{
    return [f = std::move(f)](auto... args) {
        return f(std::forward<decltype(args)>(args)...); 
    };
}

template<typename T, typename F = bool(const T&, const T&)>
inline auto compare_func(F f)
{
    return select_func(std::move(f));
}
```
which can be used like: 
```c++
bool my_compare(int, int);
bool my_compare(const public_key&, const public_key&);
std::sort(vec1.begin(), vec1.end(), select_func<bool(int, int)>(my_compare));
std::sort(vec2.begin(), vec2.end(), compare_func<public_key>(my_compare));
```

Reply:
I wrote a tool that works for my use case. You're kind of saying I shouldn't have this tool because it doesn't work for code I didn't write... (comparison functions that aren't using `const&` parameters).

In practice, user-defined comparison free-functions should almost always be using `const&` parameters. Trivial types don't need a user-defined function - just use the built-in `operator<`. If anything, enforcing `const&` prevents accidentally writing expensive copies.

For completeness I added your `select_func` thing (renamed to `as_functor`).

Reply:
> return [f = std::move(f)](auto... args) {
>        return f(std::forward<decltype(args)>(args)...); 
>    };

Btw, don't the `args` here become value parameters (copies), not references?

Reply:
Yup, I botched that particular portion, it needs to be `auto&&...`.